[FEA] Add DELTA_BINARY_PACKED decoding support to Parquet reader

etseidl opened the pull request:
## Description
Part of #13501. This adds support for decoding Parquet pages that are DELTA_BINARY_PACKED.

In addition to adding delta support, this PR incorporates changes introduced in #13622, such as using a mask to determine which decoding kernels to run, and adding parameters to  the `page_state_buffers_s` struct to reduce the amount of shared memory used.

<!-- Provide a standalone description of changes in this PR. -->
<!-- Reference any issues closed by this PR with "closes #1234". -->
<!-- Note: The pull request title will be included in the CHANGELOG. -->

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

vuule commented:
/ok to test

etseidl commented on cpp/src/io/parquet/page_string_utils.cuh:
```diff
@@ -0,0 +1,110 @@
+  }
+}
+
+/**
```
change comment to match page_string_decode.cu

vuule commented:
/ok to test

vuule commented:
/ok to test

vuule commented:
/ok to test

vuule reviewed (commented):
Partial review; nothing major so far

etseidl commented on cpp/src/io/parquet/rle_stream.cuh:
```diff
@@ -144,8 +139,19 @@ struct rle_run {
-template <typename level_t>
+template <typename level_t, int decode_threads>
 struct rle_stream {
+  // TODO: consider if these should be template parameters to rle_stream
```
Copy/paste from #13622, but I see your point.

etseidl commented on cpp/src/io/parquet/parquet_gpu.hpp:
```diff
@@ -48,6 +48,12 @@ constexpr size_type MAX_DICT_SIZE = (1 << MAX_DICT_BITS) - 1;
 // level decode buffer size.
 constexpr int LEVEL_DECODE_BUF_SIZE = 2048;
 
+template <int rolling_size>
+constexpr int rolling_index(int index)
+{
+  return index % rolling_size;
+}
+
```
Another bit stolen from @nvdbaranec 😅

vuule replied:
Regardless, I like that it's reused in quite a few places now:)
I do wonder if having `rolling_size` as a template parameter is actually better than simply passing it as a second parameter.

etseidl replied:
I would think having it as a parameter would allow the compiler to better optimize the mod

etseidl commented on cpp/src/io/parquet/page_data.cu:
```diff
@@ -908,36 +590,19 @@ __global__ void __launch_bounds__(decode_block_size) gpuDecodePageData(
-                      bool compute_string_sizes,
-                      int level_type_size,
-                      rmm::cuda_stream_view stream)
+uint32_t GetKernelMasks(cudf::detail::hostdevice_vector<PageInfo>& pages,
```
will change

galipremsagar commented on python/cudf/cudf/tests/test_parquet.py:
```diff
@@ -1306,6 +1306,38 @@ def test_parquet_reader_v2(tmpdir, simple_pdf):
+def test_delta_binary(add_nulls, tmpdir):
+    nrows = 100000
+    # Create a pandas dataframe with random data of mixed types
+    test_pdf = pd.DataFrame(
+        {
+            "col_int32": pd.Series(
+                np.random.randint(0, 0x7FFFFFFF, nrows), dtype="Int32"
+            ),
+            "col_int64": pd.Series(
+                np.random.randint(0, 0x7FFFFFFFFFFFFFFF, nrows), dtype="Int64"
+            ),
+        },
+    )
```
We can do this:
```python
In [9]: from cudf.testing.dataset_generator import rand_dataframe

In [10]: rand_dataframe(
    ...:         dtypes_meta=[
    ...:             {"dtype": "int32", "null_frequency": 0, "cardinality": 10},
    ...:             {"dtype": "int64", "null_frequency": 0, "cardinality": 10},
    ...:         ],
    ...:         rows=10,
    ...:         use_threads=False,
    ...:     )
Out[10]: 
pyarrow.Table
0: int32 not null
1: int64 not null
----
0: [[-960623239,-960623239,-960623239,1802422255,813242668,1533446499,-1866640613,813242668,-960623239,1052894290]]
1: [[3688646814997525975,9072133112761886084,71084674236880465,71084674236880465,-3180049500932746967,1310175474504604399,-2548288514416915645,-2548288514416915645,-3180049500932746967,7296846495910963387]]

In [11]: arrow_table = rand_dataframe(
    ...:         dtypes_meta=[
    ...:             {"dtype": "int32", "null_frequency": 0, "cardinality": 10},
    ...:             {"dtype": "int64", "null_frequency": 0, "cardinality": 10},
    ...:         ],
    ...:         rows=10,
    ...:         use_threads=False,
    ...:     )

In [12]: arrow_table.to_pandas()
Out[12]: 
            0                    1
0  -960623239  3688646814997525975
1  -960623239  9072133112761886084
2  -960623239    71084674236880465
3  1802422255    71084674236880465
4   813242668 -3180049500932746967
5  1533446499  1310175474504604399
6 -1866640613 -2548288514416915645
7   813242668 -2548288514416915645
8  -960623239 -3180049500932746967
9  1052894290  7296846495910963387

In [13]: arrow_table.to_pandas().dtypes
Out[13]: 
0    int32
1    int64
dtype: object

In [14]: test_pdf = arrow_table.to_pandas()

In [15]: test_pdf.columns = ['col_int32', 'col_int64']

In [16]: test_pdf
Out[16]: 
    col_int32            col_int64
0  -960623239  3688646814997525975
1  -960623239  9072133112761886084
2  -960623239    71084674236880465
3  1802422255    71084674236880465
4   813242668 -3180049500932746967
5  1533446499  1310175474504604399
6 -1866640613 -2548288514416915645
7   813242668 -2548288514416915645
8  -960623239 -3180049500932746967
9  1052894290  7296846495910963387

In [17]: test_pdf.dtypes
Out[17]: 
col_int32    int32
col_int64    int64
dtype: object
```

etseidl replied:
I'm remembering why I did it this way now...this was the only way I could figure out to get nulls into the mix without turning the datatypes into floats.  If I use @galipremsagar's code snippet with `null_frequency` non-zero, I get the wrong data types.
```python
>>> arrow_table = dg.rand_dataframe(
...     dtypes_meta=[
...             {"dtype": "int32", "null_frequency": 0.25, "cardinality": 10},
...             {"dtype": "int64", "null_frequency": 0.25, "cardinality": 10},
...     ],
...     rows=10,
...     use_threads=False,
... )
>>> pdf=arrow_table.to_pandas()
>>> pdf
              0             1
0 -1.559097e+09  1.899959e+18
1  8.516883e+08 -3.530339e+18
2           NaN  8.763204e+18
3           NaN -5.233053e+18
4           NaN  8.763204e+18
5           NaN           NaN
6  2.066054e+09 -9.136197e+18
7 -4.568663e+08  8.554710e+18
8           NaN  8.763204e+18
9 -1.897442e+08  4.816579e+17
>>> pdf.dtypes
0    float64
1    float64
dtype: object
```

galipremsagar replied:
Ah, sorry to mislead you. I wasn't conscious that you were looking for nullable dataframe's. That's an issue with Arrow's `to_pandas`. You could read it directly to cudf and do `to_pandas(nullable=True)`:
```python
In [7]: arrow_table = rand_dataframe(
   ...:         dtypes_meta=[
   ...:             {"dtype": "int32", "null_frequency": 0.23, "cardinality": 10},
   ...:             {"dtype": "int64", "null_frequency": 0.5, "cardinality": 10},
   ...:         ],
   ...:         rows=10,
   ...:         use_threads=False,
   ...:     )

In [8]: arrow_table
Out[8]: 
pyarrow.Table
0: int32
1: int64
----
0: [[1332042624,96165984,709323938,-2007798989,66183341,1332042624,null,709323938,null,-720495091]]
1: [[null,null,-1112871718206661933,null,null,null,-1678709838643542400,-2116964278366894809,-3980802044486337838,7904556893162515738]]

In [9]: cudf.DataFrame.from_arrow(arrow_table)
Out[9]: 
             0                     1
0   1332042624                  <NA>
1     96165984                  <NA>
2    709323938  -1112871718206661933
3  -2007798989                  <NA>
4     66183341                  <NA>
5   1332042624                  <NA>
6         <NA>  -1678709838643542400
7    709323938  -2116964278366894809
8         <NA>  -3980802044486337838
9   -720495091   7904556893162515738

In [10]: cudf.DataFrame.from_arrow(arrow_table).dtypes
Out[10]: 
0    int32
1    int64
dtype: object

In [11]: df = cudf.DataFrame.from_arrow(arrow_table)

In [12]: df.to_pandas(nullable=True)
Out[12]: 
             0                     1
0   1332042624                  <NA>
1     96165984                  <NA>
2    709323938  -1112871718206661933
3  -2007798989                  <NA>
4     66183341                  <NA>
5   1332042624                  <NA>
6         <NA>  -1678709838643542400
7    709323938  -2116964278366894809
8         <NA>  -3980802044486337838
9   -720495091   7904556893162515738

In [13]: df.to_pandas(nullable=True).dtypes
Out[13]: 
0    Int32
1    Int64
dtype: object
```

etseidl replied:
Thanks! I'll give that a whirl.

vuule commented:
/ok to test

vuule commented on cpp/src/io/parquet/page_delta_decode.cu:
```diff
@@ -0,0 +1,164 @@
+  auto* const sb        = &state_buffers;
+  int const page_idx    = blockIdx.x;
+  int const t           = threadIdx.x;
+  int const lane_id     = t & 0x1f;
```
```suggestion
  int const lane_id     = t & cudf::detail::warp_size;
```

vuule reviewed (commented):
Went over the rest of the changes :)
Bunch of small questions/suggestions.

etseidl commented on cpp/src/io/parquet/delta_binary.cuh:
```diff
@@ -0,0 +1,281 @@
+  inline __device__ void decode_batch()
+  {
+    int const t       = threadIdx.x;
+    int const lane_id = t & 0x1f;
```
you mean `t % cudf::detail::warp_size` right?

etseidl commented on cpp/src/io/parquet/delta_binary.cuh:
```diff
@@ -0,0 +1,281 @@
+
+  // skip to the start of the next mini-block. should only be called on thread 0.
+  // calls init_binary_block if currently on the last mini-block in a block.
+  inline __device__ void setup_next_mini_block(bool is_decode)
```
yeah, sorry...artifact of splitting the delta decoders into two PRs.  It will eventually be called with `is_decode == false`

vuule replied:
Not an issue. Glad I guessed correctly :)

etseidl commented on cpp/src/io/parquet/delta_binary.cuh:
```diff
@@ -0,0 +1,281 @@
+
+      // unpack deltas. modified from version in gpuDecodeDictionaryIndices(), but
+      // that one only unpacks up to bitwidths of 24. simplified some since this
+      // will always do batches of 32. also replaced branching with a loop.
```
I'll have to dig up old commits.  I used to have the loop unrolled...but didn't notice a huge difference at the time.  I'll revisit.

etseidl replied:
I tried replacing the loop with my old unrolled version:
```c++
          if (c < mb_bits && p < block_end) {  // up to 8 bits
            delta |= static_cast<zigzag128_t>(*p++) << c;
            c += 8;
            if (c < mb_bits && p < block_end) {  // up to 16 bits
              delta |= static_cast<zigzag128_t>(*p++) << c;
              c += 8;
              if (c < mb_bits && p < block_end) {  // 24 bits
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                if (c < mb_bits && p < block_end) {  // 32 bits
                  delta |= static_cast<zigzag128_t>(*p++) << c;
                  c += 8;
                  if (c < mb_bits && p < block_end) {  // 40 bits
                    delta |= static_cast<zigzag128_t>(*p++) << c;
                    c += 8;
                    if (c < mb_bits && p < block_end) {  // 48 bits
                      delta |= static_cast<zigzag128_t>(*p++) << c;
                      c += 8;
                      if (c < mb_bits && p < block_end) {  // 56 bits
                        delta |= static_cast<zigzag128_t>(*p++) << c;
                        c += 8;
                        if (c < mb_bits && p < block_end) {  // 64 bits
                          delta |= static_cast<zigzag128_t>(*p++) << c;
                        }
                      }
                    }
                  }
                }
              }
            }
          }
```
I also tried a version that uses a switch rather than nested branching:
```c++
          auto const bytes_left = (mb_bits - ofs + 7) / 8;
          if (p + bytes_left <= block_end) {
            switch (bytes_left) {
              case 8:
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                [[fallthrough]];
              case 7:
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                [[fallthrough]];
              case 6:
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                [[fallthrough]];
              case 5:
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                [[fallthrough]];
              case 4:
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                [[fallthrough]];
              case 3:
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                [[fallthrough]];
              case 2:
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                [[fallthrough]];
              case 1: delta |= static_cast<zigzag128_t>(*p++) << c; [[fallthrough]];
              default: break;
            }
          }
```
I ran each of those through a 10M line file with columns of random int32 and int64. The int64 deltas should get pretty large and exercise the deeper parts of the unpacker.  In the end, the `while` and unrolled `if`s were pretty much a wash, as I remembered.  The switch version was a bit slower :(

vuule replied:
omg, that code. Good to see that the loop is no worse than other options.
Thanks for checking this.

etseidl replied:
do you think the unrolled is worth the register savings (48 vs 55)?

vuule replied:
I don't think it is. Feels like a "gaze into the abyss" sort of ordeal :D
Could be persuaded otherwise with a benchmark, though :shrug:

etseidl replied:
ok, I'll leave it alone then.  will just add a note.

etseidl commented on cpp/src/io/parquet/page_delta_decode.cu:
```diff
@@ -0,0 +1,164 @@
+                                int level_type_size,
+                                rmm::cuda_stream_view stream)
+{
+  CUDF_EXPECTS(pages.size() > 0, "There is no page to decode");
```
copied that from `DecodePageData`...would have to change it there as well.

vuule replied:
Let's re-imagine this as a TODO :)
What would be your preference?
I'm bringing this up because if we introduce a bug where this is called w/o pages in some corner case, a user would just see `There is no page to decode`, which is not very instructive. Instead, the reader could just work (correctly) :man_shrugging:

etseidl replied:
Yeah, I've always been kind of puzzled by that test anyway.  I could at least change the error message to say no delta pages. Let me think on this.

etseidl replied:
So there's an identical check in `reader::impl::decode_page_data()` (also in `reader::impl::read_chunk_internal()`), so the checks in the decode methods are both safe and unnecessary. I vote leave them alone for now.

etseidl commented on cpp/src/io/parquet/page_delta_decode.cu:
```diff
@@ -0,0 +1,164 @@
+
+    if (t < 64) {  // warp0..1
+      target_pos = min(src_pos + 2 * batch_size, s->nz_count + batch_size);
+    } else {       // warp2...
```
yes...did you mean to flag line 106? ` } else if (t < 96 && src_pos < target_pos) {`

vuule replied:
I'm asking because of the comment; IMO `...` implies more warps after warp2.
I guess line 106 also implies more threads after 95. Is this future-proofing?
Sorry for being nitpicky here, just trying to help future readers (which is probably us :D).

etseidl replied:
bad eyes...didn't see the elipsis.  will smack.

vuule commented:
/ok to test

vuule commented on cpp/src/io/parquet/page_decode.cuh:
```diff
@@ -954,21 +956,11 @@ struct all_types_filter {
-  __device__ inline bool operator()(PageInfo const& page) { return is_string_col(page, chunks); }
+struct mask_filter {
+  int mask;
+  __device__ inline bool operator()(PageInfo const& page) { return (page.kernel_mask & mask) != 0; }
```
`setupLocalPageInfo` docs below are a bit broken - return value description is missing, as well as description of the last parameter. Not a request for this PR, more of a note.

etseidl replied:
There was a lot of docstring cruft in this file, so I cleaned up most of it.

vuule reviewed (approved):
Looks great, thanks for addressing the comments!

etseidl commented:
> Looks great, thanks for addressing the comments!

Thanks for the reviews! Good thing I started on the encoder...that helped root out a few more bugs. 😅

vuule commented:
/ok to test

nvdbaranec commented on cpp/src/io/parquet/rle_stream.cuh:
```diff
@@ -22,16 +22,11 @@
-constexpr int num_rle_stream_decode_warps =
-  (num_rle_stream_decode_threads / cudf::detail::warp_size) - 1;
-constexpr int run_buffer_size = (num_rle_stream_decode_warps * 2);
-constexpr int rolling_run_index(int index) { return index % run_buffer_size; }
```
Glad to see this being unified with `rolling_index`.

nvdbaranec commented on cpp/src/io/parquet/parquet_gpu.hpp:
```diff
@@ -458,6 +477,16 @@ void BuildStringDictionaryIndex(ColumnChunkDesc* chunks,
+ * @param[in] stream CUDA stream to use
+ * @return Bitwise OR of all page `kernel_mask` values
+ */
+uint32_t SumPageKernelMasks(cudf::detail::hostdevice_vector<PageInfo>& pages,
```
This function name feels a little oblique.  I think `GetKernelMask` would be cleaner, with a description like "Determine the set of kernels that need to be invoked on these pages."

vuule replied:
My concern with `GetKernelMask` is that it looks like a getter. If it were something like `GetAggregatedKernelMask` it would be more clear to me that some reduction is happening in the functions.

etseidl replied:
Also coming down the pike is encode kernel masks...how about `GetAggregateDecodeKernelMasks`?

nvdbaranec replied:
Yeah, that sounds good to me.

nvdbaranec commented on cpp/src/io/parquet/page_string_utils.cuh:
```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2018-2023, NVIDIA CORPORATION.
```
```suggestion
 * Copyright (c) 2023, NVIDIA CORPORATION.
```

nvdbaranec commented on cpp/src/io/parquet/decode_preprocess.cu:
```diff
@@ -0,0 +1,417 @@
+/*
+ * Copyright (c) 2018-2023, NVIDIA CORPORATION.
```
```suggestion
 * Copyright (c) 2023, NVIDIA CORPORATION.
```

nvdbaranec commented on cpp/src/io/parquet/parquet_gpu.hpp:
```diff
@@ -553,6 +582,25 @@ void DecodeStringPageData(cudf::detail::hostdevice_vector<PageInfo>& pages,
+ * @param[in] chunks All chunks to be decoded
+ * @param[in] num_rows Total number of rows to read
+ * @param[in] min_row Minimum number of rows to read
+ * @param[in] stream CUDA stream to use, default 0
```
missing param for `level_type_size`

etseidl replied:
fixed

nvdbaranec reviewed (changes requested):
Quick first pass. More to come.

etseidl commented on cpp/src/io/parquet/rle_stream.cuh:
```diff
@@ -22,16 +22,11 @@
-constexpr int rolling_run_index(int index) { return index % run_buffer_size; }
+constexpr int rle_stream_required_run_buffer_size(int num_threads)
+{
+  int num_rle_stream_decode_warps = (num_threads / cudf::detail::warp_size) - 1;
```
seems the compiler isn't smart enough to figure out that the passed in `num_threads` is known at compile time. Had to pass `num_threads` as a template param to get this working.

nvdbaranec commented on cpp/src/io/parquet/page_string_utils.cuh:
```diff
@@ -0,0 +1,110 @@
+  // `dst + 4` and `len - 4` are used instead of `dst` and `len` to avoid
+  // `load_uint4` reading beyond string boundaries.
+  // use signed int since out_end_aligned can be negative.
+  int64_t out_start_aligned = (in_datatype_size + alignment_offset + out_datatype_size - 1) /
```
```suggestion
  int64_t const out_start_aligned = (in_datatype_size + alignment_offset + out_datatype_size - 1) /
```

nvdbaranec commented on cpp/src/io/parquet/page_string_utils.cuh:
```diff
@@ -0,0 +1,110 @@
+  int64_t out_start_aligned = (in_datatype_size + alignment_offset + out_datatype_size - 1) /
+                                out_datatype_size * out_datatype_size -
+                              alignment_offset;
+  int64_t out_end_aligned =
```
```suggestion
  int64_t const out_end_aligned =
```

bdice reviewed (approved):
CMake approval.

galipremsagar commented on python/cudf/cudf/tests/test_parquet.py:
```diff
@@ -1286,6 +1286,55 @@ def test_parquet_reader_v2(tmpdir, simple_pdf):
+                "cardinality": nrows,
+            },
+        ],
+        rows=nrows,
```
```suggestion
        rows=nrows,
        seed=0,
```

galipremsagar replied:
We can pass the seed here for determinism.

galipremsagar reviewed (approved):
Approving with a suggestion.

vuule commented:
/ok to test

vuule commented:
/merge
